### PR TITLE
add search user, converstaion name, and string methods. Started integ…

### PR DESCRIPTION
**Pull request by xujy**
Created search functions in chat.java to search for string in message, search for author of message, and search conversations. Also started integrating db class with model so there is persistence with the users now.

### CODE REVIEW

File: make.sh [hunk 1]
```diff
@@ -16,7 +16,12 @@
 
 set -e
 
+
+cd ./database/
+sqlite3 main.db < build.sql
```
Comment (gnarph) [inline make.sh]: Could you remove both cd lines and use
`sqlite3 database/min.db < database/build.sql

File: src/codeu/chat/client/ClientMessage.java [hunk 2]
```diff
@@ -126,6 +128,49 @@ public void showAllMessages() {
     }
   }
 
+  // For m-search-string command.
+  // Show all messages attached to the current conversation containing the string keyword
+  public void searchString(String key) {
+    boolean found = false;
+    if (conversationContents.size() == 0) {
+      System.out.println(" Current Conversation has no messages");
```
Comment (gnarph) [inline src/codeu/chat/client/ClientMessage.java]: Extra space at the beginning of the string.

File: src/codeu/chat/client/ClientMessage.java [hunk 3]
```diff
@@ -126,6 +128,49 @@ public void showAllMessages() {
     }
   }
 
+  // For m-search-string command.
+  // Show all messages attached to the current conversation containing the string keyword
+  public void searchString(String key) {
+    boolean found = false;
+    if (conversationContents.size() == 0) {
```
Comment (gnarph) [inline src/codeu/chat/client/ClientMessage.java]: Do we really need this as an extra case?
If we do, just return from this block and don't use an else

File: src/codeu/chat/client/ClientMessage.java [hunk 4]
```diff
@@ -126,6 +128,49 @@ public void showAllMessages() {
...
+  }
+
+
+  // For m-search-author command.
+  // Show all messages attached to the current conversation by a certain author
+  public void searchAuthor(String author) {
+    boolean found = false;
+
+    if (conversationContents.size() == 0) {
+      System.out.println(" Current Conversation has no messages");
```
Comment (gnarph) [inline src/codeu/chat/client/ClientMessage.java]: Same comment from above applies here

File: src/codeu/chat/client/ClientMessage.java [hunk 5]
```diff
@@ -126,6 +128,49 @@ public void showAllMessages() {
...
+    } else {
+      for (final Message m : conversationContents) {
+
```
Comment (gnarph) [inline src/codeu/chat/client/ClientMessage.java]: No reason to have an extra newline here.

File: src/codeu/chat/client/ClientMessage.java [hunk 6]
```diff
@@ -126,6 +128,49 @@ public void showAllMessages() {
...
+    if (!found) {
+      System.out.println(" No match found for " + key);
+    }
+  }
+
+
+  // For m-search-author command.
+  // Show all messages attached to the current conversation by a certain author
+  public void searchAuthor(String author) {
```
Comment (gnarph) [inline src/codeu/chat/client/ClientMessage.java]: This function is much the same as searchString, could they share code?

Comment (gnarph) [inline src/codeu/chat/client/ClientMessage.java]: I think having these functions return a list of the applicable authors/messages/etc would be easier to test and use than the current version that only prints. If we return data instead, the caller can decide what to do with it; printing is rather limited.

File: src/codeu/chat/client/commandline/Chat.java [hunk 7]
```diff
@@ -150,6 +154,16 @@ private void doOneCommand(Scanner lineScanner) {
 
       selectConversation(lineScanner);
 
+    } else if (token.equals("c-search-title")) {
+
+      if (!tokenScanner.hasNext()) {
+        System.out.println("ERROR: Conversation title not supplied.");
```
Comment (gnarph) [inline src/codeu/chat/client/commandline/Chat.java]: Should we just print out an error? Would it make more sense to raise an exception so the code can deal with what has happened? Is this the right place to decide how a lack of title should be handled?

File: src/codeu/chat/client/commandline/Chat.java [hunk 8]
```diff
@@ -165,7 +179,7 @@ private void doOneCommand(Scanner lineScanner) {
               clientContext.conversation.getCurrentId(),
               body);
 
-            respondAsBot(body);
+          /*respondAsBot(body);*/
```
Comment (gnarph) [inline src/codeu/chat/client/commandline/Chat.java]: Remove this line if we're not going to use it.

File: src/codeu/chat/client/commandline/Chat.java [hunk 9]
```diff
@@ -335,7 +390,7 @@ public void selectConversation(Scanner lineScanner) {
     }
   }
 
-  private void respondAsBot(String body) {
+  /*rivate void respondAsBot(String body) {
```
Comment (gnarph) [inline src/codeu/chat/client/commandline/Chat.java]: Delete commented blocks, git keeps their history if we ever need them back.

File: src/codeu/chat/server/Model.java [hunk 10]
```diff
@@ -74,10 +77,13 @@ public void add(User user) {
...
     userByTime.insert(user.creation, user);
     userByText.insert(user.name, user);
+
+    db.insertUser(user);
   }
 
   public StoreAccessor<Uuid, User> userById() {
-    return userById;
+    return db.listUsersId();
+    //return userById;
```
Comment (gnarph) [inline src/codeu/chat/server/Model.java]: Delete commented.

File: src/codeu/chat/server/Model.java [hunk 11]
```diff
@@ -25,9 +25,12 @@
 import codeu.chat.util.Uuid;
 import codeu.chat.util.store.Store;
 import codeu.chat.util.store.StoreAccessor;
+import codeu.chat.util.dbUtils;
```
Comment (gnarph) [inline src/codeu/chat/server/Model.java]: To conform with the rest of the style, this should start with a capital (DbUtils).

Classes starting with capitals makes it easier to tells objects apart from class types.

File: src/codeu/chat/util/dbUtils.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+import java.sql.*;
+
+import java.util.Comparator;
+import codeu.chat.common.User;
+import codeu.chat.common.Conversation;
+import codeu.chat.common.Message;
+import codeu.chat.util.store.Store;
+
+public class dbUtils {
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Change name please (see other comment).

File: src/codeu/chat/util/dbUtils.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+                return 1;
+            }
+
+            final int order = Integer.compare(a.id(), b.id());
+            return order == 0 ? compare(a.root(), b.root()) : order;
+        }
+    };
+
+    private Connection connect() {
+        // SQLite connection string
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Comments should tell us "why" more than "what"; a reader can probably figure out what the string is. If not the variable name should make it clear (perhaps sql_connection_string is clearer than url).

File: src/codeu/chat/util/dbUtils.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        String url = "jdbc:sqlite:database/main.db";
+        Connection conn = null;
+        try {
+            conn = DriverManager.getConnection(url);
+        } catch (SQLException e) {
+            logger.error("Connection to database failed.", e.getMessage());
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: This swallows an exception that could lead to the caller getting a null connection. Perhaps we should create our own database error class and raise that when there are problems?

File: src/codeu/chat/util/dbUtils.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+        String sql = "INSERT INTO users(user_id, name, creation) VALUES(?,?,?)";
+        try (Connection conn = this.connect();
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
+            pstmt.setInt(1, user.id.id());
+            pstmt.setString(2, user.name);
+            pstmt.setLong(3, user.creation.inMs());
+            pstmt.executeUpdate();
+            logger.info("User insertion successful.");
+        } catch (SQLException e) {
+            logger.error(e.getMessage());
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Similar to above; we should tell the caller somehow when we were unable to insert a user (probably an exception).

File: src/codeu/chat/util/dbUtils.java [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+        try (Connection conn = this.connect();
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
+            pstmt.setInt(1, conversation.id.id());
+            pstmt.setString(2, conversation.owner.toString());
+            pstmt.setLong(3, conversation.creation.inMs());
+            pstmt.setString(4, conversation.title);
+            pstmt.executeUpdate();
+            logger.info("Conversation insertion successful");
+        } catch (SQLException e) {
+            logger.error(e.getMessage());
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Ditto

File: src/codeu/chat/util/dbUtils.java [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+            pstmt.setInt(1, message.id.id());
+            pstmt.setInt(2, message.author.id());
+            pstmt.setInt(3, message.next.id());
+            pstmt.setInt(4, message.previous.id());
+            pstmt.setLong(5,  message.creation.inMs());
+            pstmt.setString(6, message.content);
+            pstmt.executeUpdate();
+            logger.info("Message insertion successful.");
+        } catch (SQLException e) {
+            logger.error(e.getMessage());
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Ditto

File: src/codeu/chat/util/dbUtils.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+        }
+    }
+
+
+
+    public Store<Uuid, User> listUsersId(){
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Please add documentation (javadoc or other comments) for public methods.

This applies to the other functions too.

File: src/codeu/chat/util/dbUtils.java [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+        String sql = "SELECT user_id, name, creation FROM users";
+        final Store<Uuid, User> userById = new Store<>(UUID_COMPARE);
+
+        try (Connection conn = this.connect();
+             Statement stmt = conn.createStatement();
+             ResultSet rs = stmt.executeQuery(sql)) {
+
+            while (rs.next()) {
+                User user = new User(new Uuid(rs.getInt("user_id")), rs.getString("name"), Time.fromMs(rs.getLong("creation")));
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Please make this line shorter (and others). 100 characters is a good upper bound for java.

File: src/codeu/chat/util/dbUtils.java [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+            }
+
+        } catch (SQLException e) {
+            logger.error(e.getMessage());
+        }
+
+        return messageById;
+    }
+
+    /*public void listConversations(){
```
Comment (gnarph) [inline src/codeu/chat/util/dbUtils.java]: Delete commented blocks.

File: make.sh (hunk 1 above)
Comment (xujy) [inline make.sh]: resolved

File: src/codeu/chat/client/ClientMessage.java (hunk 3 above)
Comment (xujy) [inline src/codeu/chat/client/ClientMessage.java]: resolved

File: src/codeu/chat/client/ClientMessage.java (hunk 2 above)
Comment (xujy) [inline src/codeu/chat/client/ClientMessage.java]: resolved

File: src/codeu/chat/client/ClientMessage.java (hunk 5 above)
Comment (xujy) [inline src/codeu/chat/client/ClientMessage.java]: resolved

File: src/codeu/chat/client/commandline/Chat.java (hunk 8 above)
Comment (xujy) [inline src/codeu/chat/client/commandline/Chat.java]: resolved

File: src/codeu/chat/client/commandline/Chat.java (hunk 9 above)
Comment (xujy) [inline src/codeu/chat/client/commandline/Chat.java]: resolved

File: src/codeu/chat/server/Model.java (hunk 11 above)
Comment (xujy) [inline src/codeu/chat/server/Model.java]: resolved

File: src/codeu/chat/server/Model.java (hunk 10 above)
Comment (xujy) [inline src/codeu/chat/server/Model.java]: resolved

File: src/codeu/chat/util/dbUtils.java (hunk 12 above)
Comment (xujy) [inline src/codeu/chat/util/dbUtils.java]: resolved

File: src/codeu/chat/util/dbUtils.java (hunk 13 above)
Comment (xujy) [inline src/codeu/chat/util/dbUtils.java]: resolved

File: src/codeu/chat/util/dbUtils.java (hunk 19 above)
Comment (xujy) [inline src/codeu/chat/util/dbUtils.java]: resolved

File: src/codeu/chat/util/dbUtils.java (hunk 20 above)
Comment (xujy) [inline src/codeu/chat/util/dbUtils.java]: resolved

File: src/codeu/chat/util/dbUtils.java (hunk 18 above)
Comment (xujy) [inline src/codeu/chat/util/dbUtils.java]: resolved

Comment (xujy) [inline src/codeu/chat/util/dbUtils.java]: what should be included in documentation? talk about what the method does? or how it is called?

File: src/codeu/chat/util/dbUtils.java (hunk 15 above)
Comment (xujy) [inline src/codeu/chat/util/dbUtils.java]: Will keep working on this

**Comment by xujy** [comment]
accidentally merged this.. but i reverted it

**Pull request merged**